api/src/main/java/jakarta/config/TypeToken.java

```diff
@@ -0,0 +1,403 @@
+        // …
+        // The erasure of a nested type T.C is
+        // |T|.C. [Class.getDeclaringClass() returns an already erased
+        // type.]
+        //
+        // The erasure of an array type T[] is |T|[]. [A Class that is
+        // an array has a Class as its component type, and that is
+        // already erased.]
+        // …
+        // The erasure of every other type is the type itself. [So in all
+        // cases we can just return the supplied Class<?>.]
+        return type;
+    }
+
+    private static final Class<?> erase(ParameterizedType type) {
+        // https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.6
+        // …
+        // The erasure of a parameterized type (§4.5) G<T1,…,Tₙ> is
+        // |G| [|G| means the erasure of G, i.e. the erasure of
+        // type.getRawType()].
+        return erase(type.getRawType());
+    }
+
+    private static final Class<?> erase(GenericArrayType type) {
+        // https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.6
+        //
+        // The erasure of an array type T[] is |T|[]. [|T| means the
+        // erasure of T. We erase the genericComponentType() and use
+        // Class#arrayType() to find the "normal" array class for the
+        // erasure.]
+        Class<?> componentType = erase(type.getGenericComponentType());
+        if (componentType == null) {
+            return null;
+        } else {
+            // Needs Java 17
+            // return componentType.arrayType();
+
+            // (Java 17's Class#arrayType() does exactly this behind the scenes.)
+            return Array.newInstance(componentType, 0).getClass();
+        }
+    }
+
+    private static final Class<?> erase(TypeVariable<?> type) {
+        // https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.6
+        //
+        // The erasure of a type variable (§4.4) is the erasure of its
+        // leftmost bound. [In the case of a TypeVariable<?> that
+        // returns multiple bounds, we know they will start with a
+        // class, not an interface and not a type variable.]
+        Type[] bounds = type.getBounds();
+        return bounds.length > 0 ? erase(bounds[0]) : Object.class;
+    }
+
+    private static final Class<?> erase(WildcardType type) {
+        // https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.6
+        //
+        // The erasure of a type variable (§4.4) is the erasure of its
+        // leftmost bound.  [WildcardTypes aren't really in the JLS
+        // per se but they behave like type variables. Only upper
+        // bounds will matter here.]
+        Type[] bounds = type.getUpperBounds();
+        return bounds != null && bounds.length > 0 ? erase(bounds[0]) : Object.class;
+    }
+
+
+    /*
+     * Static methods.
+     */
+
+
+    private static final ParameterizedType mostSpecializedParameterizedSuperclass(Type type) {
```

Comment:
e.g. my `TypeToken` subclass actually extends `YourTypeToken` subclass which extends `HerTypeToken` subclass which extends `TypeToken`; we want the actual type arguments from the most specialized of these (mine)